python_scripts/linear_models_sol_03.py

```diff
@@ -55,42 +69,65 @@
     linear_regression,
     data,
     target,
-    scoring="neg_mean_absolute_error",
-    return_estimator=True,
     cv=10,
+    scoring="neg_mean_absolute_error",
     n_jobs=2,
 )
 
 # %% [markdown]
-# Compute the mean and std of the MAE in thousands of dollars (k$).
+# Compute the mean and std of the MAE in grams (g).
 
 # %%
 # solution
 print(
-    "Mean absolute error on testing set: "
-    f"{-cv_results['test_score'].mean():.3f} k$ ± "
-    f"{cv_results['test_score'].std():.3f}"
+    "Mean absolute error on testing set with original features: "
+    f"{-cv_results['test_score'].mean():.3f} ± "
+    f"{cv_results['test_score'].std():.3f} g"
 )
 
 # %% [markdown]
-# Inspect the fitted model using a box plot to show the distribution of values
-# for the coefficients returned from the cross-validation. Hint: use the
-# function
-# [`df.plot.box()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.box.html)
-# to create a box plot.
-
+# Now create a pipeline using `make_pipeline` consisting of a
+# `PolynomialFeatures` and a linear regression. Set `degree=2` and
+# `interaction_only=True` to the feature engineering step. Remember not to
+# include the bias to avoid redundancies with the linear's regression intercept.
+#
+# Use the same strategy as before to cross-validate such pipeline.
 
 # %%
 # solution
-import pandas as pd
+from sklearn.preprocessing import PolynomialFeatures
+from sklearn.pipeline import make_pipeline
 
-weights = pd.DataFrame(
-    [est.coef_ for est in cv_results["estimator"]], columns=data.columns
+poly_features = PolynomialFeatures(
+    degree=2, include_bias=False, interaction_only=True
+)
+linear_regression_interactions = make_pipeline(
+    poly_features, linear_regression
 )
 
-# %% tags=["solution"]
-import matplotlib.pyplot as plt
+cv_results = cross_validate(
+    linear_regression_interactions,
+    data,
+    target,
+    cv=10,
+    scoring="neg_mean_absolute_error",
+    n_jobs=2,
+)
+
+# %% [markdown]
+# Compute the mean and std of the MAE in grams (g) and compare with the results
+# without feature engineering.
+
+# %%
+# solution
+print(
+    "Mean absolute error on testing set with interactions: "
+    f"{-cv_results['test_score'].mean():.3f} ± "
+    f"{cv_results['test_score'].std():.3f} g"
+)
 
-color = {"whiskers": "black", "medians": "black", "caps": "black"}
-weights.plot.box(color=color, vert=False)
-_ = plt.title("Value of linear regression coefficients")
+# %% [markdown] tags=["solution"]
+# We can observe that the mean absolute error is lower and less spread with the
+# enriched features. In this case the "interactions" are indeed predictive. In
+# the following notebook we will see what happens when the enriched features are
+# non-predictive and how to deal with this case.
```

Comment:
We could add another question here:

Write code to answer the following question: how many features are generated at the output of the `PolynomialFeatures` feature transformer in the previous pipeline?

In addition, use set_output("pandas") to transform the first 5 rows of the dataset and look at the column names. Check that the values for the new interactions features are correct for a few of them.

Reply:
Then we could further expand this exercise by asking:

> Now let's try to see if we can build an alternatively pipeline with
> fewer intermediate features while keeping a similar predictive power.
> To this end let's try to use the `Nystroem` transformer instead of
> `PolynomialFeatures`. Set the `kernel` parameter to `"poly"` and
> `degree` to `2`. Adjust the number of `components` to be as small as
> possible while keeping a good cross-validation performance.

Reply:
Then in the solution we could say that `Nystroem` is a nice alternative to `PolynomialFeatures` that makes it possible to keep the memory usage of the transformed dataset under control. However, interpreting the meaning of the intermediate features can be challenging.